src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -446,8 +496,8 @@ private bool TryInsert(TKey key, TValue value, InsertionBehavior behavior)
             IEqualityComparer<TKey>? comparer = _comparer;
             uint hashCode = (uint)((comparer == null) ? key.GetHashCode() : comparer.GetHashCode(key));
 
-            int collisionCount = 0;
-            ref int bucket = ref _buckets[hashCode % (uint)_buckets.Length];
+            int allowedCollisions = entries.Length;
```

Comment:
This adds extra memory access to the common path and makes the less common path smaller.

It may be useful to look at this change in isolation to validate that it is really a net improvement.

Reply:
I was hoping the Jit would optimize the dec to zero path better; but it adds an extra `test` https://github.com/dotnet/coreclr/pull/27149#issuecomment-541442795. It might drop the `test` if its `cmp` as will see if can get them both in registers (have decreased the register pressure as part of this change so I think everything is now).

Reply:
i.e. I might get the advantages I was looking for using the previous approach and dropping the early access.

Reply:
Works nicely https://github.com/dotnet/coreclr/pull/27149/commits/2b2cc4764a049b32fa2f8db48b25c241d74d1f33
```asm
       dec      r13d
       test     r13d, r13d
       jl       SHORT G_M59265_IG19
       jmp      SHORT G_M59265_IG12
```